README.md
# ejson
An smart json,similar to the jquery, read and update the json data.

一个小巧的、完全无依赖的类似jquery写法的读取和更改json的工具，可以按照json的层级或者某个属性读取json数据。

## Installing aelog
- ```pip install eqjson```

## Usage
假如json数据如下：
```
[
   {
      'object':{
         'name':'myLittleObject',
         'value':[
            {
               'key2':'value2',
               'key3':'value3',
               'key1':'value1'
            },
            {
               'key33':'value33',
               'key1':'value11'
            },
            {
               'key333':'value333',
               'key1':'value111'
            }
         ],
         'type':'list',
         'valueAsString':[
            'one',
            'two'
         ]
      }
   }
]
```
```
from eqjson import EasyQueryjson

# 实例化
ejson_obj = EasyQueryjson(json_doc)
# 查询
ejson_obj.get_value("object.name")
ejson_obj.get_value("object.valueAsString.0")
# update
ejson_obj.change_value("object.name", "updatename")
ejson_obj.change_value("object.value.2", ["five", "four"])
# add
ejson_obj.append_value("object.valueAsString", "three")
ejson_obj.append_value("object.valueA", {"test":456})
# remove
ejson_obj.remove_value("object.valueAsString")
ejson_obj.remove_value("object.value")

# 具体的查询参数详情如下
```
#### 基本用法：
- 如果要获取name值：则写法为```0.object.name```
- 如果json最外层是{}获取name值：则写法为```object.name```
- 如果要获取valueAsString中的第一个值：则写法为```0.object.valueAsString.0```
- 如果json最外层是{}获取valueAsString中的第一个值：则写法为```object.valueAsString.0```
- 获取value列表中的第一个值的key1值：则写法为```0.object.value.0.key1```
- 获取value列表中的所有的key1值：则写法为```0.object.value.key1```， 则结果为```['value1','value11', 'value111']```
#### 高级用法：
- 支持通过属性确定值，如果json是无序的，要获取key1值为value1所在节点的key3的值，则写法为```0.object.value[key1=value1].key3```
- 支持通过多属性确定值，比如要获取key1值为value1，key2值为value2，所在节点的key3的值，则写法为```0.object.value[key1=value1，key2=value2].key3```
- 支持多层属性，比如要获取name为myLittleObject的以上的值：则写法为```[name=myLittleObject].object.value[key1=value1].key3```


## 单测
- 100%文件覆盖率，92%的行覆盖率


CHANGES.md
## Eqjson Changelog

###[1.0.0] - 2018-03-30

#### Added 

- 增加eqjson对json字符串的普通和高级用法
- 可以通过json层级获取值
- 可以通过类似jquery的属性获取值


#### Changed


tests/__init__.py
#!/usr/bin/env python3
# coding=utf-8

"""
@author: guoyanfeng
@software: PyCharm
@time: 17-7-28 上午10:45
"""



tests/test_ejson.py
#!/usr/bin/env python3
# coding=utf-8

"""
@author: guoyanfeng
@software: PyCharm
@time: 2017-4-13 18:57
"""
import json
import unittest

from eqjson import EasyQueryjson


class TestEjson1(unittest.TestCase):
    """
    ejson 单测第一种情况，以大括号开始
    """

    @classmethod
    def setUpClass(cls):
        with open("json_test_1.json") as f:
            json_doc = json.load(f)
        cls.ejson_obj = EasyQueryjson(json_doc)

    def test_101_get_first_tier(self, ):
        """

        Args:
            self,
        Returns:

        """
        val = self.ejson_obj.get_value("id")
        self.assertEqual(val, 1123123812831823)

    def test_102_get_second_tier(self, ):
        """

        Args:
            self,
        Returns:

        """
        val = self.ejson_obj.get_value("object.name")
        self.assertEqual(val, "myLittleObject")

    def test_103_get_third_tier(self, ):
        """

        Args:
            self,
        Returns:

        """
        val = self.ejson_obj.get_value("object.valueAsString.0")
        self.assertEqual(val, "one")

    def test_104_get_third_tier_list0_dict_value(self, ):
        """

        Args:
            self,
        Returns:

        """
        val = self.ejson_obj.get_value("object.value.1.key")
        self.assertEqual(val, "value2")

    def test_105_get_list_all_key_value(self, ):
        """

        Args:
            self,
        Returns:

        """
        val = self.ejson_obj.get_value("object.value.key")
        self.assertEqual(val, ["value1", "value2", "value3"])

    def test_106_change_first_tier(self, ):
        """

        Args:
            self,
        Returns:

        """
        self.ejson_obj.change_value("id", "123456")
        val = self.ejson_obj.get_value("id")
        self.assertEqual(val, 123456)

    def test_107_change_second_tier(self, ):
        """

        Args:
            self,
        Returns:

        """
        self.ejson_obj.change_value("object.name", "updatename")
        val = self.ejson_obj.get_value("object.name")
        self.assertEqual(val, "updatename")

    def test_108_change_list_value(self, ):
        """

        Args:
            self,
        Returns:

        """
        self.ejson_obj.change_value("object.value.2", ["five", "four"])
        val = self.ejson_obj.get_value("object.value.2")
        self.assertEqual(val, ["five", "four"])

    def test_111_append_value_list(self, ):
        """

        Args:
            self,
        Returns:

        """
        self.ejson_obj.append_value("object.value", "append_test")
        val = self.ejson_obj.get_value("object.value")
        self.assertIn("append_test", val)

    def test_112_append_value_dict(self, ):
        """

        Args:
            self,
        Returns:

        """
        self.ejson_obj.append_value("object", {"test_name": "append_test"})
        val = self.ejson_obj.get_value("object.test_name")
        self.assertEqual(val, "append_test")

    def test_113_remove_list_data_value(self, ):
        """

        Args:
            self,
        Returns:

        """
        self.ejson_obj.remove_value("object.value.0")
        val = self.ejson_obj.get_value("object.value")
        self.assertEqual(len(val), 3)

    def test_114_remove_list_value(self, ):
        """

        Args:
            self,
        Returns:

        """
        self.ejson_obj.remove_value("object.value")
        val = self.ejson_obj.get_value("object.value")
        self.assertIsNone(val)


class TestEjson2(unittest.TestCase):
    """
    ejson 单测第二种情况，以列表开始
    """

    @classmethod
    def setUpClass(cls):
        with open("json_test_2.json") as f:
            json_doc = json.load(f)
        cls.ejson_obj = EasyQueryjson(json_doc)

    def test_101_get_first_tier(self, ):
        """

        Args:
            self,
        Returns:

        """
        val = self.ejson_obj.get_value("1.id")
        self.assertEqual(val, 123456)

    def test_102_get_second_tier(self, ):
        """

        Args:
            self,
        Returns:

        """
        val = self.ejson_obj.get_value("0.object.name")
        self.assertEqual(val, "myLittleObject")

    def test_103_get_third_tier(self, ):
        """

        Args:
            self,
        Returns:

        """
        val = self.ejson_obj.get_value("1.object.valueAsString.0")
        self.assertEqual(val, "one")

    def test_104_get_third_tier_list0_dict_value(self, ):
        """

        Args:
            self,
        Returns:

        """
        val = self.ejson_obj.get_value("1.object.value.1.key")
        self.assertEqual(val, "value2")

    def test_105_get_list_all_key_value(self, ):
        """

        Args:
            self,
        Returns:

        """
        val = self.ejson_obj.get_value("0.object.value.key")
        self.assertEqual(val, ["value1", "value2", "value3"])

    def test_106_change_first_tier(self, ):
        """

        Args:
            self,
        Returns:

        """
        self.ejson_obj.change_value("0.id", "123456")
        val = self.ejson_obj.get_value("0.id")
        self.assertEqual(val, 123456)

    def test_107_change_second_tier(self, ):
        """

        Args:
            self,
        Returns:

        """
        self.ejson_obj.change_value("0.object.name", "updatename")
        val = self.ejson_obj.get_value("0.object.name")
        self.assertEqual(val, "updatename")

    def test_107_change_list_value(self, ):
        """

        Args:
            self,
        Returns:

        """
        self.ejson_obj.change_value("1.object.value.2", ["five", "four"])
        val = self.ejson_obj.get_value("1.object.value.2")
        self.assertEqual(val, ["five", "four"])


class TestEjson3(unittest.TestCase):
    """
    ejson 单测第二种情况，以列表开始
    """

    @classmethod
    def setUpClass(cls):
        with open("json_test_3.json") as f:
            json_doc = json.load(f)
        cls.ejson_obj = EasyQueryjson(json_doc)

    def test_101_get_first_tier(self, ):
        """

        Args:
            self,
        Returns:

        """
        val = self.ejson_obj.get_value("[id=123456]")
        self.assertEqual(val["id"], 123456)

    def test_102_get_second_tier(self, ):
        """

        Args:
            self,
        Returns:

        """
        val = self.ejson_obj.get_value("[id=123456].object.value")
        self.assertEqual(len(val), 3)

    def test_103_get_third_tier(self, ):
        """

        Args:
            self,
        Returns:

        """
        val = self.ejson_obj.get_value("[id=123456].object.value[key1=value1]")
        self.assertEqual(len(val), 3)

    def test_104_get_third_tier_list0_dict_value(self, ):
        """

        Args:
            self,
        Returns:

        """
        val = self.ejson_obj.get_value("[id=123456].object.value[key1=value1].key3")
        self.assertEqual(val, "value3")

    def test_105_get_list_all_key_value(self, ):
        """

        Args:
            self,
        Returns:

        """
        val = self.ejson_obj.get_value("1.object.value[key11=value11].key33")
        self.assertEqual(val, "value33")

    def test_106_get_value_by_mul_attr(self, ):
        """

        Args:
            self,
        Returns:

        """
        val = self.ejson_obj.get_value("1.object.value[key11=value11,key22=value22, key33=value33].key33")
        self.assertEqual(val, "value33")

    def test_107_get_list_by_mul_attr(self, ):
        """

        Args:
            self,
        Returns:

        """
        val = self.ejson_obj.get_value("1.object.value[key11=value11,key22=value22, key33=value33]")
        self.assertEqual(len(val), 3)

    def test_108_not_found_by_mul_attr(self, ):
        """

        Args:
            self,
        Returns:

        """
        val = self.ejson_obj.get_value("1.object.value[key11=value11,key22=value2, key33=value33]")
        self.assertIs(val, None)


if __name__ == '__main__':
    # unittest.main(verbosity=2)
    test_load = unittest.TestLoader()
    suite = test_load.loadTestsFromTestCase(TestEjson1)
    suite.addTest(test_load.loadTestsFromTestCase(TestEjson2))
    suite.addTest(test_load.loadTestsFromTestCase(TestEjson3))
    unittest.TextTestRunner(verbosity=2).run(suite)


eqjson/eqjson.py
#!/usr/bin/env python3
# coding=utf-8

"""
@author: guoyanfeng
@software: PyCharm
@time: 18-3-30 下午1:32
"""

__all__ = ["EasyQueryjson"]


class EasyQueryjson(object):
    """
    通用的字符串语法类获取、更改、添加json的值

    暂时不支持根节点的添加操作
    """

    def __init__(self, json_obj):
        """
        通用的字符串语法类获取、更改、添加json的值

        暂时不支持根节点的添加操作
        Args:
            json_obj:python对象,如果是json则需要load后再传参
        Returns:

        """
        self.json_obj = json_obj

    def get_value(self, path):
        """
        通过通用的path寻址来获得对应的值
        Args:
            path
        Returns:

        """
        if "[" in path and "]" in path:
            return self._get_value_by_attr(path)
        else:
            return self._get_value(path)

    def change_value(self, path, value):
        """
        通过path和value来改变path对应的值
        Args:
            path
            value

        Returns:

        """
        json_value = self.get_value(path)
        cutted_path = path.rsplit('.', 1)[0]
        parent_json_value = self.get_value(cutted_path)

        # 对更改json的值做校验，防止改错
        if isinstance(parent_json_value, dict) or path == cutted_path:
            if self._is_value(json_value) and self._is_value(value):
                self._set_value(path, self._change_value_type(json_value, value))
            elif isinstance(value, list) and isinstance(json_value, list):
                self._set_value(path, value)
            elif isinstance(value, dict) and isinstance(json_value, dict):
                self._set_value(path, value)
            else:
                raise ValueError("更改的值{0}和模板中的值{1}的类型不一致,请检查!".format(type(value), type(json_value)))
        elif isinstance(parent_json_value, list):
            self._set_value(path, value)

    def remove_value(self, path):
        """
        通过path删除对应json节点的值
        Args:
            path

        Returns:

        """
        self._set_value(path, None)

    def append_value(self, path, value):
        """
        通过path和value来添加path对应的值
        Args:
            path
            value

        Returns:

        """
        json_value = self.get_value(path)
        if isinstance(json_value, list):
            json_value.append(value)
        elif isinstance(json_value, dict):
            self._append_entry(json_value, value)
        else:
            raise ValueError("基本类型值{0} {1}无法进行添加操作".format(json_value, type(json_value)))

    def _get_value(self, path, json_obj=None):
        """
        通过通用的path寻址来获得对应的值
        Args:
            path:
            json_obj: 提供的json doc 或者 self.json_obj
        Returns:

        """
        if path:
            splitted_path = str(path).split('.')
            return self._get_value_by_path(splitted_path, json_obj or self.json_obj)
        else:
            return json_obj or self.json_obj

    @staticmethod
    def _append_entry(json_value, value):
        """
        更新字典
        Args:
            json_value
            value

        Returns:

        """
        if isinstance(value, dict):
            json_value.update(value)
        else:
            raise ValueError("添加的值{0}和模板中的值{1}的类型不一致".format(type(value), type(json_value)))

    @staticmethod
    def _change_value_type(src_value, value):
        """
        把value的类型尽量更改为和src_value一样
        Args:
            src_value, value
        Returns:

        """
        try:
            return type(src_value)(value)
        except ValueError:
            return value

    def _set_value(self, path, value):
        """
        更改值的实际方法,更改值测试过了，删除值没有测试过
        Args:
            path
            value

        Returns:

        """
        if path.count('.') >= 1:
            cutted_path = path.rsplit('.', 1)
            obj = self.get_value(cutted_path[0])
            if isinstance(obj, dict):
                if value is None:
                    del obj[cutted_path[1]]
                else:
                    obj[cutted_path[1]] = value
            elif isinstance(obj, list):
                if cutted_path[1].isdigit():
                    if value is None:
                        obj.pop(int(cutted_path[1]))
                    else:
                        obj[int(cutted_path[1])] = value
                else:
                    raise ValueError("{0}类型是list，而{1}类型不是number，无法取值".format(type(obj), type(cutted_path[1])))

        else:
            if isinstance(self.json_obj, dict):
                if value is None:
                    del self.json_obj[path]
                else:
                    self.json_obj[path] = value
            elif isinstance(self.json_obj, list):
                if path.isdigit():
                    if value is None:
                        self.json_obj.pop(int(path))
                    else:
                        self.json_obj[int(path)] = value
                else:
                    raise ValueError("{0}类型是list，而{1}类型不是number，无法取值".format(type(self), type(path)))

    def _get_value_by_path(self, path, json_obj):
        """
        递归调用来取得对应path的值
        Args:
            path
            json_obj

        Returns:

        """
        if len(path) == 0:
            return json_obj
        e = path[0]
        json_obj = self._get_one_level_deeper(e, json_obj)
        path.remove(e)
        return self._get_value_by_path(path, json_obj)

    def _get_one_level_deeper(self, input_key, json_obj):
        """
        判断各种类型，通过inputkey，获得json_obj对应的值,支持json是字典或者是列表,列表直接通过数字访问
        Args:
            input_key:
            json_obj

        Returns:

        """

        if isinstance(json_obj, dict):
            return json_obj.get(input_key)
        elif isinstance(json_obj, list):
            if self._is_list_of_objects(json_obj):
                if input_key.isdigit():
                    # 如果是数字直接返回对应的值
                    return json_obj[int(input_key)]
                else:
                    # 如果不是数字返回列表中每项的key对应的值组成的列表
                    return self._get_values_from_list_by_key(input_key, json_obj)
            elif self._is_list_of_not_dict_objects(json_obj):
                if input_key.isdigit():
                    return json_obj[int(input_key)]
                else:
                    return json_obj
            else:
                return json_obj

    @staticmethod
    def _is_list_of_objects(input_list):
        """
        判断值得类型，判断是否是列表并且包含字典
        Args:
            input_list

        Returns:

        """
        if isinstance(input_list, list) and len(input_list) > 0:
            return True if isinstance(input_list[0], dict) else False

    def _is_list_of_not_dict_objects(self, input_list):
        """
        判断值得类型，判断是否是列表并且包含除字典外的类型
        Args:
            input_list

        Returns:

        """
        return not self._is_list_of_objects(input_list)

    @staticmethod
    def _is_value(value):
        return True if isinstance(value, (int, str, float, complex)) else False

    @staticmethod
    def _get_values_from_list_by_key(key, json_obj):
        """
        如果jsonobj是一个列表，那么返回类表中对应的每一个key的值
        Args:
            key:
            json_obj

        Returns:

        """
        result = []
        for e in json_obj:
            if isinstance(e, dict):
                result.append(e.get(key))
        return result

    def _get_val_from_unordered_list_by_unique_val(self, json_obj, path):
        """
        从无序的列表中，通过一个唯一值或者多个唯一值得到改列表中确定的节点
        Args:
            json_obj:需要查找的python对象
            path: 需要获取值的path路径，eg data.name[id=123]，data.name[id=123,path=test]
        Returns:
            返回由data.name[id=123]或者data.name[id=123,path=test]确定的节点
        """
        path = path.replace(" ", "")
        pre_path = path[:path.index("[")]
        pre_node_val = self._get_value(pre_path, json_obj)

        if isinstance(pre_node_val, list):
            attr_path = path[path.index("[") + 1:path.index("]")]
            if "," in attr_path:
                attr_pairs_value = [val.split("=") for val in attr_path.split(",")]
            else:
                attr_pairs_value = [attr_path.split("=")]
            for node in pre_node_val:
                for attr_name, attr_value in attr_pairs_value:
                    if str(node.get(attr_name)) != attr_value:
                        break
                else:
                    return node
            else:
                return None
        else:
            return pre_node_val

    def _get_value_by_attr(self, path):
        """
        通过path中的属性获取由属性确定的节点，目前每层节点能指定一个或多个属性，可以在多层节点上指定属性
        Args:
            path：需要获取值的path路径，eg data.name[id=123, path=test].name[id=1234].name
        Returns:
            返回由data.name[id=123, path=test].name[id=1234].name确定的值
        """
        path = path.replace(" ", "")
        json_obj = self.json_obj
        if path.count("]") == 1:
            left_path = path[:path.index("]") + 1]
            after_path = path[path.index("]") + 1:].strip(".")
            json_obj = self._get_val_from_unordered_list_by_unique_val(json_obj, left_path)
            return self._get_value(after_path, json_obj) if after_path else json_obj
        else:
            after_path = path
            for _ in range(after_path.count("]")):
                left_path = after_path[:after_path.index("]") + 1]
                json_obj = self._get_val_from_unordered_list_by_unique_val(json_obj, left_path)
                after_path = after_path[after_path.index("]") + 1:].strip(".")
            return self._get_value(after_path, json_obj) if after_path else json_obj


ejson/ejson.py
#!/usr/bin/env python3
# coding=utf-8

"""
@author: guoyanfeng
@software: PyCharm
@time: 18-3-30 下午1:32
"""

__all__ = ["Ejson"]


class Ejson(object):
    """
    通用的字符串语法类获取、更改、添加json的值

    暂时不支持根节点的添加操作
    """

    def __init__(self, json_obj):
        """
        通用的字符串语法类获取、更改、添加json的值

        暂时不支持根节点的添加操作
        Args:
            json_obj:python对象,如果是json则需要load后再传参
        Returns:

        """
        self.json_obj = json_obj

    def _get_value_from_path(self, path, json_obj=None):
        """
        通过通用的path寻址来获得对应的值
        Args:
            path:
            json_obj: 提供的json doc 或者 self.json_obj
        Returns:

        """
        if path:
            splitted_path = str(path).split('.')
            return self._get_value_by_path(splitted_path, json_obj or self.json_obj)
        else:
            return json_obj or self.json_obj

    @staticmethod
    def _append_entry(json_value, value):
        """
        更新字典
        Args:
            json_value
            value

        Returns:

        """
        if isinstance(value, dict):
            json_value.update(value)
        else:
            raise ValueError("添加的值{0}和模板中的值{1}的类型不一致".format(type(value), type(json_value)))

    def append_value(self, path, value):
        """
        通过path和value来添加path对应的值
        Args:
            path
            value

        Returns:

        """
        json_value = self.get_value_from_path(path)
        if isinstance(json_value, list):
            json_value.append(value)
        elif isinstance(json_value, dict):
            self._append_entry(json_value, value)
        else:
            raise ValueError("基本类型值{0} {1}无法进行添加操作".format(json_value, type(json_value)))

    @staticmethod
    def _change_value_type(src_value, value):
        """
        把value的类型尽量更改为和src_value一样
        Args:
            src_value, value
        Returns:

        """
        try:
            return type(src_value)(value)
        except ValueError:
            return value

    def change_value(self, path, value):
        """
        通过path和value来改变path对应的值
        Args:
            path
            value

        Returns:

        """
        json_value = self.get_value_from_path(path)
        cutted_path = path.rsplit('.', 1)[0]
        parent_json_value = self.get_value_from_path(cutted_path)

        # 对更改json的值做校验，防止改错
        if isinstance(parent_json_value, dict) or path == cutted_path:
            if self._is_value(json_value) and self._is_value(value):
                self._set_value(path, self._change_value_type(json_value, value))
            elif isinstance(value, list) and isinstance(json_value, list):
                self._set_value(path, value)
            elif isinstance(value, dict) and isinstance(json_value, dict):
                self._set_value(path, value)
            else:
                raise ValueError("更改的值{0}和模板中的值{1}的类型不一致,请检查!".format(type(value), type(json_value)))
        elif isinstance(parent_json_value, list):
            self._set_value(path, value)

    def remove_value(self, path):
        """
        通过path删除对应json节点的值
        Args:
            path

        Returns:

        """
        self._set_value(path, None)

    def _set_value(self, path, value):
        """
        更改值的实际方法,更改值测试过了，删除值没有测试过
        Args:
            path
            value

        Returns:

        """
        if path.count('.') >= 1:
            cutted_path = path.rsplit('.', 1)
            obj = self.get_value_from_path(cutted_path[0])
            if isinstance(obj, dict):
                if value is None:
                    del obj[cutted_path[1]]
                else:
                    obj[cutted_path[1]] = value
            elif isinstance(obj, list):
                if cutted_path[1].isdigit():
                    if value is None:
                        obj.pop(int(cutted_path[1]))
                    else:
                        obj[int(cutted_path[1])] = value
                else:
                    raise ValueError("{0}类型是list，而{1}类型不是number，无法取值".format(type(obj), type(cutted_path[1])))

        else:
            if isinstance(self.json_obj, dict):
                if value is None:
                    del self.json_obj[path]
                else:
                    self.json_obj[path] = value
            elif isinstance(self.json_obj, list):
                if path.isdigit():
                    if value is None:
                        self.json_obj.pop(int(path))
                    else:
                        self.json_obj[int(path)] = value
                else:
                    raise ValueError("{0}类型是list，而{1}类型不是number，无法取值".format(type(self), type(path)))

    def _get_value_by_path(self, path, json_obj):
        """
        递归调用来取得对应path的值
        Args:
            path
            json_obj

        Returns:

        """
        if len(path) == 0:
            return json_obj
        e = path[0]
        json_obj = self._get_one_level_deeper(e, json_obj)
        path.remove(e)
        return self._get_value_by_path(path, json_obj)

    def _get_one_level_deeper(self, input_key, json_obj):
        """
        判断各种类型，通过inputkey，获得json_obj对应的值,支持json是字典或者是列表,列表直接通过数字访问
        Args:
            input_key:
            json_obj

        Returns:

        """

        if isinstance(json_obj, dict):
            return json_obj.get(input_key)
        elif isinstance(json_obj, list):
            if self._is_list_of_objects(json_obj):
                if input_key.isdigit():
                    # 如果是数字直接返回对应的值
                    return json_obj[int(input_key)]
                else:
                    # 如果不是数字返回列表中每项的key对应的值组成的列表
                    return self._get_values_from_list_by_key(input_key, json_obj)
            elif self._is_list_of_not_dict_objects(json_obj):
                if input_key.isdigit():
                    return json_obj[int(input_key)]
                else:
                    return json_obj
            else:
                return json_obj

    @staticmethod
    def _is_list_of_objects(input_list):
        """
        判断值得类型，判断是否是列表并且包含字典
        Args:
            input_list

        Returns:

        """
        if isinstance(input_list, list) and len(input_list) > 0:
            return True if isinstance(input_list[0], dict) else False

    def _is_list_of_not_dict_objects(self, input_list):
        """
        判断值得类型，判断是否是列表并且包含除字典外的类型
        Args:
            input_list

        Returns:

        """
        return not self._is_list_of_objects(input_list)

    @staticmethod
    def _is_value(value):
        return True if isinstance(value, (int, str, float, complex)) else False

    @staticmethod
    def _get_values_from_list_by_key(key, json_obj):
        """
        如果jsonobj是一个列表，那么返回类表中对应的每一个key的值
        Args:
            key:
            json_obj

        Returns:

        """
        result = []
        for e in json_obj:
            if isinstance(e, dict):
                result.append(e.get(key))
        return result

    def _get_val_from_unordered_list_by_unique_val(self, json_obj, path):
        """
        从无序的列表中，通过一个唯一值或者多个唯一值得到改列表中确定的节点
        Args:
            json_obj:需要查找的python对象
            path: 需要获取值的path路径，eg data.name[id=123]，data.name[id=123,path=test]
        Returns:
            返回由data.name[id=123]或者data.name[id=123,path=test]确定的节点
        """
        path = path.replace(" ", "")
        pre_path = path[:path.index("[")]
        pre_node_val = self._get_value_from_path(pre_path, json_obj)

        if isinstance(pre_node_val, list):
            attr_path = path[path.index("[") + 1:path.index("]")]
            if "," in attr_path:
                attr_pairs_value = [val.split("=") for val in attr_path.split(",")]
            else:
                attr_pairs_value = [attr_path.split("=")]
            for node in pre_node_val:
                for attr_name, attr_value in attr_pairs_value:
                    if str(node.get(attr_name)) != attr_value:
                        break
                else:
                    return node
            else:
                return None
        else:
            return pre_node_val

    def _get_value_from_path_by_attr(self, path):
        """
        通过path中的属性获取由属性确定的节点，目前每层节点能指定一个或多个属性，可以在多层节点上指定属性
        Args:
            path：需要获取值的path路径，eg data.name[id=123, path=test].name[id=1234].name
        Returns:
            返回由data.name[id=123, path=test].name[id=1234].name确定的值
        """
        path = path.replace(" ", "")
        json_obj = self.json_obj
        if path.count("]") == 1:
            left_path = path[:path.index("]") + 1]
            after_path = path[path.index("]") + 1:].strip(".")
            json_obj = self._get_val_from_unordered_list_by_unique_val(json_obj, left_path)
            return self._get_value_from_path(after_path, json_obj) if after_path else json_obj
        else:
            after_path = path
            for _ in range(after_path.count("]")):
                left_path = after_path[:after_path.index("]") + 1]
                json_obj = self._get_val_from_unordered_list_by_unique_val(json_obj, left_path)
                after_path = after_path[after_path.index("]") + 1:].strip(".")
            return self._get_value_from_path(after_path, json_obj) if after_path else json_obj

    def get_value_from_path(self, path):
        """
        通过通用的path寻址来获得对应的值
        Args:
            path
        Returns:

        """
        if "[" in path and "]" in path:
            return self._get_value_from_path_by_attr(path)
        else:
            return self._get_value_from_path(path)
